src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -286,52 +476,119 @@ static size_t uct_cuda_copy_md_get_total_device_mem(CUdevice cuda_device)
     return 1; /* return 1 byte to avoid division by zero */
 }
 
+static ucs_status_t
+uct_cuda_copy_detect_vmm(const void *address, CUmemorytype *cuda_mem_type,
+                         unsigned *is_vmm_allocation)
+{
+    ucs_status_t status = UCS_OK;
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop = {};
+    CUmemGenericAllocationHandle alloc_handle;
+
+    /* Check if memory is allocated using VMM API and see if host memory needs
+     * to be treated as pinned device memory */
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemRetainAllocationHandle(&alloc_handle, (void*)address));
+    if (status != UCS_OK) {
+        status = UCS_OK;
+        goto ret;
+    }
+
+    *is_vmm_allocation = 1;
+    if (*cuda_mem_type != CU_MEMORYTYPE_HOST) {
+        goto mem_release;
+    }
+
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(
+            cuMemGetAllocationPropertiesFromHandle(&prop, alloc_handle));
+    if (status != UCS_OK) {
+        status = UCS_ERR_INVALID_ADDR;
+        goto mem_release;
+    }
+
+    /* if location type is HOST the memory can be potentially accessed by CEs */
+    if ((prop.location.type == CU_MEM_LOCATION_TYPE_HOST) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA) ||
+        (prop.location.type == CU_MEM_LOCATION_TYPE_HOST_NUMA_CURRENT)) {
+        *cuda_mem_type = CU_MEMORYTYPE_DEVICE;
+    } else {
+        status = UCS_ERR_INVALID_ADDR;
+    }
+
+mem_release:
+    status = UCT_CUDADRV_FUNC_LOG_DEBUG(cuMemRelease(alloc_handle));
+ret:
+#endif
+    return status;
+}
+
+static void
+uct_cuda_copy_set_sync_memops(const void *address, unsigned is_vmm_allocation)
+{
+    unsigned value = 1;
+
+#if HAVE_CUDA_FABRIC
```

Comment:
is this `#if` really needed?
is it ok to set SYNC_MEMOPS per context (globally) while this function accepts a single address? if yes why do it for every buffer and not once during initialization?

Reply:
1. `#if` is not really needed. Suffices to check if address is vmm and set memops for the whole context.
2. It's ok to do it and should be done once per context. Setting this context isn't expensive so was setting each time but I'll change that.